plugins/woocommerce/src/Internal/ProductQueryFilters/FilterDataProvider.php

```diff
@@ -0,0 +1,364 @@
+
+		$this->set_cache( $transient_key, $results );
+
+		return $results;
+	}
+
+	/**
+	 * Get attribute counts for the current products.
+	 *
+	 * @param array  $query_vars         The WP_Query arguments.
+	 * @param string $attribute_to_count Attribute taxonomy name.
+	 * @return array termId=>count pairs.
+	 */
+	public function get_attribute_counts( $query_vars, $attribute_to_count ) {
+		$pre_filter_data = $this->pre_get_filter_data( $query_vars, 'attribute', array( 'taxonomy' => $attribute_to_count ) );
+
+		if ( ! empty( $pre_filter_data ) ) {
+			return $pre_filter_data;
+		}
+
+		$transient_key = $this->get_transient_key( $query_vars, 'attribute', array( 'taxonomy' => $attribute_to_count ) );
+		$cached_data   = $this->get_cache( $transient_key );
+
+		if ( ! empty( $cached_data ) ) {
+			return $cached_data;
+		}
+
+		global $wpdb;
+
+		add_filter( 'posts_clauses', array( $this->filter_clauses_generator, 'add_query_clauses' ), 10, 2 );
+		add_filter( 'posts_pre_query', '__return_empty_array' );
+
+		$query_vars['no_found_rows']  = true;
+		$query_vars['posts_per_page'] = -1;
+		$query_vars['fields']         = 'ids';
+		$query                        = new \WP_Query();
+
+		$query->query( $query_vars );
+		$product_query_sql = $query->request;
+
+		remove_filter( 'posts_clauses', array( $this->filter_clauses_generator, 'add_query_clauses' ), 10 );
+		remove_filter( 'posts_pre_query', '__return_empty_array' );
+
+		$attributes_to_count_sql = 'AND term_taxonomy.taxonomy IN ("' . esc_sql( wc_sanitize_taxonomy_name( $attribute_to_count ) ) . '")';
+		$attribute_count_sql     = "
+			SELECT COUNT( DISTINCT posts.ID ) as term_count, terms.term_id as term_count_id
+			FROM {$wpdb->posts} AS posts
+			INNER JOIN {$wpdb->term_relationships} AS term_relationships ON posts.ID = term_relationships.object_id
+			INNER JOIN {$wpdb->term_taxonomy} AS term_taxonomy USING( term_taxonomy_id )
+			INNER JOIN {$wpdb->terms} AS terms USING( term_id )
+			WHERE posts.ID IN ( {$product_query_sql} )
+			{$attributes_to_count_sql}
+			GROUP BY terms.term_id
+		";
+
+		$results = $wpdb->get_results( $attribute_count_sql ); // phpcs:ignore
+		$results = array_map( 'absint', wp_list_pluck( $results, 'term_count', 'term_count_id' ) );
+
+		$this->set_cache( $transient_key, $results );
+
+		return $results;
+	}
+
+	/**
+	 * Generate calculate query by stock status.
+	 *
+	 * @param string $status status to calculate.
+	 * @param string $product_query_sql product query for current filter state.
+	 * @param array  $stock_status_options available stock status options.
+	 *
+	 * @return false|string
+	 */
+	private function generate_stock_status_count_query( $status, $product_query_sql, $stock_status_options ) {
+		if ( ! in_array( $status, $stock_status_options, true ) ) {
+			return false;
+		}
+		global $wpdb;
+		$status = esc_sql( $status );
+		return "
+			SELECT COUNT( DISTINCT posts.ID ) as status_count
```

Comment:
You can use lookup table for this, and also a group by to compute for all stocks in the same query. stock_status is indexed so group by should be quick. Also let's cache?